lib/gamnit/tileset.nit

```diff
@@ -0,0 +1,186 @@
+	is autoinit do
+		chars_cleaned = chars.replace("\n", "")
+	end
+
+	# Character present in the texture, set by `chars=`
+	private var chars_cleaned: Text is noautoinit
+
+	# Additional space to insert horizontally between characters
+	#
+	# A negative value may display overlapped tiles.
+	var hspace: Numeric = 0.0 is writable
+
+	# Additional space to insert vertically between characters
+	#
+	# A negative value may display overlapped tiles.
+	var vspace: Numeric = 0.0 is writable
+
+	# The glyph/tile/texture associated to `char`
+	#
+	# Returns null if `char` is not in `chars`.
+	fun char(char: Char): nullable Texture
+	do
+		var i = chars_cleaned.index_of(char)
+		if i == -1 then return null
+		return subtextures[i]
+	end
+
+	# Distance between the beginning of a letter tile and the beginning of the next letter tile
+	fun advance: Float do return width.to_f + hspace.to_f
+
+	# Distance between the beginning and the end of the longest line of `text`
+	fun text_width(text: Text): Numeric
+	do
+		var lines = text.split('\n')
+		if lines.is_empty then return 0
+
+		var longest = 0
+		for line in lines do longest = longest.max(line.length)
+
+		return longest.mul(advance)
+	end
+
+	# Distance between the top of the first line to the bottom of the last line in `text`
+	fun text_height(text: Text): Numeric
+	do
+		if text.is_empty then return 0
+
+		var n_lines = text.chars.count('\n')
+		return (n_lines+1).mul(height.add(vspace)).sub(vspace)
+	end
+end
+
+# Manage a set of sprites to display some text
+class TextSprites
+
+	# Font used to draw text
+	var font: TileSetFont
+
+	# Top left of the first character in UI coordinates
+	var anchor: Point3d[Float]
+
+	# Last set of sprites generated to display `text=`
+	var sprites = new Array[Sprite]
+
+	# Sprite set where to put created sprites
+	#
+	# Defaults to `app::ui_sprites`, but it could also be set to a
+	# `app::sprites` or a custom collection.
+	var target_sprite_set: Sequence[Sprite] = app.ui_sprites is lazy, writable
+
+	private var cached_text: nullable Text = ""
+
+	# Last text drawn
+	fun text: nullable Text do return cached_text
+
+	# Update the text displayed by inserting new sprites into `app.ui_sprites`
+	#
+	# Does not redraw if `text` has not changed.
+	fun text=(text: nullable Text)
+	is autoinit do
+		# Don't redraw if text hasn't changed
+		if text == cached_text then return
+		cached_text = text
+
+		# Clean up last used sprites
+		for s in sprites do if target_sprite_set.has(s) then target_sprite_set.remove s
+		sprites.clear
+
+		if text == null then return
+
+		# Build new sprites
+		var dx = font.vspace.to_f/2.0
+		var dy = font.hspace.to_f/2.0
+		for c in text do
+			if c == '\n' then
+				dy -= font.height.to_f + font.vspace.to_f
+				dx = font.vspace.to_f/2.0
+				continue
+			else if c.is_whitespace then
+				dx += font.advance
+				continue
+			end
+
+			var tex = font.char(c)
+			if tex == null then
+				# Try to fallback to '?'
+				tex = font.char('?')
```

Comment:
A more sensible choice these days for a general fallback is � (the REPLACEMENT CHARACTER). That way, one may specify a sprite for � and it will be displayed as expected.